### PR TITLE
Correction de 1790 suspensions qui finissent dans trop longtemps

### PR DESCRIPTION
**Pull request by celine-m-s**
**Carte Notion : ** (à créer)

[Carte Notion du script précédent.](https://www.notion.so/plateforme-inclusion/2-2-Suspension-36-mois-R-gularisation-sur-le-stock-des-PASS-IAE-suspendus-pour-12-mois-BOOST-18795bec5259430382701d67a3a879e0)

### Pourquoi ?

(à rédiger)

### Comment (optionnel)

- Enregistrement des bonnes dates des suspensions en erreur.
- Gestion des exceptions: 1 PASS ayant été supprimé et 10 suspensions ayant été modifiée par un utilisateur.

### CODE REVIEW

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            "suspensions_end_at": pandas.to_datetime,
+        }
+        df = pandas.read_csv(csv_file, sep=" ", names=colnames, header=None, converters=converters)
+
+        unique_approval_numbers_from_file = set(df["approval_number"].to_list())
+        self.stdout.write(f"Total of unique approvals in input file: {len(unique_approval_numbers_from_file)}")
+
+        # Just for the records as we only have one.
+        # Should be treated manually.
+        excluded_suspensions = suspensions.exclude(approval__number__in=unique_approval_numbers_from_file)
```
Comment (rsebille) [inline itou/scripts/management/commands/update_suspensions_end_at.py:56]: _nit_: C'est du one-shot donc on s'en fiche un peu mais `suspensions_not_in_file` serait sans doute plus clair, car savoir que des suspensions sont exclus ne nous dit pas pourquoi, et obligé de lire toute la ligne pour le savoir :) #lazydev

File: itou/scripts/tests.py [hunk 2]
```diff
@@ -1,12 +1,16 @@
 import datetime
 import io
+import tempfile
```
Comment (rsebille) [inline itou/scripts/tests.py:3]: Peut-être plutôt utiliser la fixture [`tmp_path` de pytest](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) plutôt que de créer des fichiers dans le dossier temporaire de la machine locale, histoire de s'éviter des problèmes (`xdist` :eyes:).
Surtout que maintenant il y a un truc qui facilite l'utilisation des fixtures pytest pour les `TestCase` :
https://github.com/betagouv/itou/blob/ad184ed43f88c439efd87ff7732e3d4fd76f616d/itou/conftest.py#L91-L93

File: itou/scripts/tests.py [hunk 3]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+        suspension5_updated_at = update_suspensions_end_at_mngt_comd.FIRST_SCRIPT_RUNNING_DATE + relativedelta(days=1)
+        suspension5 = SuspensionFactory(
+            approval=suspension4.approval, start_at=datetime.datetime.today(), updated_at=suspension5_updated_at
+        )
+
+        with tempfile.NamedTemporaryFile() as file:
+            suspensions = Suspension.objects.prefetch_related("approval").all()
+            for suspension in suspensions.iterator():
+                log_row = f"{suspension.approval.number} {suspension.start_at} {suspension.end_at}\n"
+                file.write(bytes(log_row, encoding="utf-8"))
```
Comment (rsebille) [inline itou/scripts/tests.py:198]:
```suggestion
                file.write(log_row.encode())
```
Ou alors tu ouvres le fichier en texte et plus besoin de s’embêter normalement.

File: itou/scripts/tests.py [hunk 4]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+            # Edge cases
+            assert (
+                f"Skipping suspension updated after script ran: {suspension5.pk}, "
+                f"start_at: {suspension5.start_at.date()}, end_at: {suspension5.end_at.date()}"
+            ) in output
+            assert "Skipping approval not found: XXXXX5432764" in output
+            assert "Skipped suspensions (one suspension per approval): 1" in output
+            assert (
+                "Skipped suspensions (other reasons): 2" in output
+            )  # Approval not found, suspension updated after script ran.
```
Comment (rsebille) [inline itou/scripts/tests.py:228]: A ce compte là, je me demande si il vaudrais pas tester la sortie en entier (`assert output.getvalue().split("\n") == [...]`) plutôt que juste quelque bout ? Au moins après tu as la certitude que le test cassera à la moindre modification du script.

File: itou/scripts/tests.py [hunk 5]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+        suspension1.refresh_from_db()
+        # One suspension linked to one approval. Don't touch it.
+        assert not suspension1.updated_at
+
+        # Many suspensions linked to one approval: the end_at was updated many times.
+        # This is the first suspension. Only the last one should be updated.
+        suspension2.refresh_from_db()
+        assert not suspension2.updated_at
+        suspension3.refresh_from_db()
+        assert suspension3.updated_at
```
Comment (rsebille) [inline itou/scripts/tests.py:239]: Vu qu'il y a `freeze_gun` autant aller au bout de la logique, `updated_at` pourrais être initialisé lors de la création, surtout je doute que tout nos champs `updated_at` fonctionne de la même manière (corrigé un qui était jamais MAJ il n'y a pas longtemps :eyes:) :
```suggestion
        assert suspension3.updated_at == timezone.now()
```

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            if last_suspension.updated_at and last_suspension.updated_at > FIRST_SCRIPT_RUNNING_DATE:
+                self.stdout.write(
+                    "Skipping suspension updated after script ran: "
+                    f"{last_suspension.pk}, start_at: {last_suspension.start_at}, end_at: {last_suspension.end_at}, "
+                    f"updated_at: {last_suspension.updated_at.date()}"
+                )
+                skipped_suspensions_other_reason_counter += 1
+                continue
+
+            last_suspension.end_at -= relativedelta(months=MONTHS_OF_PROLONGATION * (row["size"] - 1))
```
Comment (rsebille) [inline itou/scripts/management/commands/update_suspensions_end_at.py:99]: J'suis toujours un peu réticent à faire de l'arithmétique plutôt que mettre une valeur absolue quand c'est possible, mais de ce que j'ai entendu de la discussion d'hier ça devrais pas poser problème dans le cas actuel.

**Review by rsebille — APPROVED** [review]
Le fond me semble OK mais c'est vu de loin, pour les remarques de formes tu en fais ce que tu veux, surtout que c'est un script one-shot ;).

File: itou/scripts/tests.py [hunk 7]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+            start_at=datetime.datetime.today() - relativedelta(days=30),
+            end_at=datetime.datetime.today() - relativedelta(days=1),
+        )
+        suspension5_updated_at = update_suspensions_end_at_mngt_comd.FIRST_SCRIPT_RUNNING_DATE + relativedelta(days=1)
+        suspension5 = SuspensionFactory(
+            approval=suspension4.approval, start_at=datetime.datetime.today(), updated_at=suspension5_updated_at
+        )
+
+        with tempfile.NamedTemporaryFile() as file:
+            suspensions = Suspension.objects.prefetch_related("approval").all()
```
Comment (francoisfreitag) [inline itou/scripts/tests.py:195]: C’est pas un `select_related` ? `Suspension` a une FK sur `Approval`.

File: itou/scripts/tests.py [hunk 8]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+            for suspension in suspensions.iterator():
+                log_row = f"{suspension.approval.number} {suspension.start_at} {suspension.end_at}\n"
+                file.write(bytes(log_row, encoding="utf-8"))
+
+                # Approval who was deleted after the first script ran.
```
Comment (francoisfreitag) [inline itou/scripts/tests.py:200]:
```suggestion
                # Approval was deleted after the first script ran.
```

File: itou/scripts/tests.py (hunk 4 above)
Comment (francoisfreitag) [reply itou/scripts/tests.py:228]: Ça garanti aussi l’ordre dans lequel les opérations s’exécute, et est plus facile à relire.

File: itou/scripts/tests.py [hunk 9]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+        # One suspension linked to one approval. Don't touch it.
+        assert not suspension1.updated_at
+
+        # Many suspensions linked to one approval: the end_at was updated many times.
+        # This is the first suspension. Only the last one should be updated.
+        suspension2.refresh_from_db()
+        assert not suspension2.updated_at
+        suspension3.refresh_from_db()
+        assert suspension3.updated_at
+        assert suspension3.end_at == (suspension3_end_at - relativedelta(months=months_of_prolongation)).date()
```
Comment (francoisfreitag) [inline itou/scripts/tests.py:240]: Puisqu’on a `freezegun`, on peut hardcoder la date attendue. C’est bien plus lisible et facile à vérifier.

**Comment by francoisfreitag** [comment]
Pas la peine de le changer, mais il aurait été plus simple de pré-traiter le CSV à la main, plutôt que de s’embêter à gérer les exceptions sur la capture de sortie du ticket dans la commande de gestion. :)

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument("--wet-run", action="store_true", dest="wet_run")
+        parser.add_argument(
+            "--file-path",
+            dest="file_path",
+            required=True,
+            action="store",
+            help="Path of the ASP CSV file to deduplicate",
+        )
```
Comment (francoisfreitag) [inline itou/scripts/management/commands/update_suspensions_end_at.py:26]: Rien d’important pour cette commande, mais comme le fichier est requis, il pourrait être en *positional argument*, plutôt qu’un *flag*.

**Review by francoisfreitag — COMMENTED** [review]
La sortie du script sur un dump récent me perturbe :

```
Problematic suspensions found in database: 1794.
Total of unique approvals in input file: 58710
Suspension not in log file. Pk: 84177, start_at: 2022-09-22, end_at: 2202-10-27
Problematic suspensions found in input file: 1793
Skipping suspension updated after script ran: 106430, start_at: 2022-11-24, end_at: 2023-02-05, updated_at: 2023-02-03
Skipping suspension updated after script ran: 123098, start_at: 2022-12-07, end_at: 2023-02-06, updated_at: 2023-02-03
Skipping suspension updated after script ran: 130757, start_at: 2023-01-05, end_at: 2024-01-04, updated_at: 2023-02-03
Skipping suspension updated after script ran: 126517, start_at: 2023-01-01, end_at: 2023-02-28, updated_at: 2023-02-03
Skipping suspension updated after script ran: 132596, start_at: 2023-01-31, end_at: 2023-02-02, updated_at: 2023-02-02
La date de fin doit être postérieure à la date de début.
>  Skipping because of validation error! Suspension: 79710, Approval number: REDACTED
start_at: 2022-01-01, end_at: 2016-12-31
Updated suspensions: 1793
Skipped suspensions (one suspension per approval): 56911
Skipped suspensions (other reasons): 6
```

Je vois que 1793 suspensions sont mises à jour, alors que 5 ont été modifiées après que le premier script ait tourné et 1 suspension est invalide, je m’attends à 1793 - 5 - 1 = 1787 suspensions modifiées ? :thinking:

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+import pandas
+from dateutil.relativedelta import relativedelta
+from django.core.exceptions import ValidationError
+from django.core.management.base import BaseCommand
+from django.utils import timezone
+
+from itou.approvals.models import Approval, Suspension
+
+
+FIRST_SCRIPT_RUNNING_DATE = timezone.make_aware(dt(2023, 2, 1, 17, 44))
```
Comment (francoisfreitag) [inline itou/scripts/management/commands/update_suspensions_end_at.py:13]: Plutôt que d’utiliser de la logique pour gérer les TZ, je recommanderai de construire directement un objet `datetime` avec `tzinfo` :

```suggestion
FIRST_SCRIPT_RUNNING_DATE = datetime.datetime(2023, 2, 1, 16, 44, tzinfo=datetime.timezone.utc)
```

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
+import io
+from datetime import datetime as dt
```
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:2]: :-D
C'est un truc que j'ai vu quand je bossais avec des data scientists et j'ai gardé cette mauvaise pratique. Je corrige.

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        # Should be treated manually.
+        excluded_suspensions = suspensions.exclude(approval__number__in=unique_approval_numbers_from_file)
+        for suspension in excluded_suspensions:
+            self.stdout.write(
+                "Suspension not in log file. "
+                f"Pk: {suspension.pk}, start_at: {suspension.start_at}, end_at: {suspension.end_at}"
+            )
+
+        suspensions = suspensions.filter(approval__number__in=unique_approval_numbers_from_file)
+        self.stdout.write(f"Problematic suspensions found in input file: {suspensions.count()}")
```
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply itou/scripts/management/commands/update_suspensions_end_at.py:64]: En fait non, on ne réutilise pas `suspensions` ailleurs.

File: itou/scripts/tests.py [hunk 14]
```diff
@@ -152,3 +156,93 @@ def test_command_output(self):
...
+        suspension3 = SuspensionFactory(
+            approval=suspension2.approval, start_at=datetime.datetime.today(), end_at=suspension3_end_at
+        )
+
+        # Suspension updated after the first script ran should not be updated.
+        suspension4 = SuspensionFactory(
+            approval__number="XXXXX0000003",
+            start_at=datetime.datetime.today() - relativedelta(days=30),
+            end_at=datetime.datetime.today() - relativedelta(days=1),
+        )
```
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/tests.py:188]: Pas sûr. Si je supprime la suspension 4, alors la suspension 5 sera considérée comme unique et passera dans le cas `if row["size"] == 1`.

File: itou/scripts/tests.py (hunk 2 above)
Comment (celine-m-s) [reply itou/scripts/tests.py:3]: Top, merci ! J'en ai profité pour utiliser Pytest (j'avais oublié... :grimacing: ).

File: itou/scripts/tests.py (hunk 4 above)
Comment (celine-m-s) [reply itou/scripts/tests.py:228]: Vous avez complètement raison. Je change.

File: itou/scripts/management/commands/update_suspensions_end_at.py (hunk 6 above)
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:99]: Hum, quelle valeur absolue voudrais-tu mettre ? Le nombre d'itérations est variable.

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        parser.add_argument(
+            "--file-path",
+            dest="file_path",
+            required=True,
+            action="store",
+            help="Path of the ASP CSV file to deduplicate",
+        )
+
+    def clean_file(self, file_path):
+        with open(file_path, newline="", encoding="utf-8") as input_file:
```
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:29]: Très juste, merci !

File: itou/scripts/management/commands/update_suspensions_end_at.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:64]: :+1:

File: itou/scripts/tests.py (hunk 7 above)
Comment (celine-m-s) [reply itou/scripts/tests.py:195]: Exact ! :)

File: itou/scripts/tests.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply itou/scripts/tests.py:188]: Haaaa, exact !

[the remark this answers is not in the document]
Comment (francoisfreitag) [reply itou/scripts/tests.py:188]: À mon avis, le test serait plus représentatif si on écrivait le CSV à la main. Après, comme c’est une commande à exécution unique, à toi de voir si le changement vaut le coup.

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        skipped_suspensions_no_duplicate_counter = 0
+        skipped_suspensions_other_reason_counter = 0
+        for _, row in df.iterrows():
+            # Only approvals with several suspensions have caused troubles.
+            if row["size"] == 1:
+                skipped_suspensions_no_duplicate_counter += 1
+                continue
+
+            try:
+                approval = Approval.objects.get(number=row["approval_number"])
```
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:82]: Tu as raison mais j'avoue avoir été paresseuse. Je change ça.

[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:82]: Après avoir un peu creusé, je n'arrive pas à remplacer très rapidement ce morceau de code par une version plus efficace :

```

            try:
                approval = Approval.objects.get(number=row["approval_number"])
            except Approval.DoesNotExist:
                self.stdout.write(f"Skipping approval not found: {row['approval_number']}")
                skipped_suspensions_other_reason_counter += 1
                continue
```
Je ne pense pas que ce soit la peine de continuer car le gain n'en vaudrait pas le temps passé.

File: itou/scripts/management/commands/update_suspensions_end_at.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+                        f">  Skipping because of validation error! Suspension: {last_suspension.pk}, "
+                        f"Approval number: {row['approval_number']}"
+                    )
+                )
+                self.stdout.write(f"start_at: {last_suspension.start_at}, end_at: {last_suspension.end_at}")
+                skipped_suspensions_other_reason_counter += 1
+                continue
+
+            if wet_run:
+                last_suspension.save(update_fields=["end_at", "updated_at"])
```
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/management/commands/update_suspensions_end_at.py:116]: Va pour la transaction !

File: itou/scripts/management/commands/update_suspensions_end_at.py (hunk 6 above)
Comment (rsebille) [reply itou/scripts/management/commands/update_suspensions_end_at.py:99]: J'avais en tête `last_suspension.end_at = last_suspension.start_at + relativedelta(months=Suspension.MAX_DURATION_MONTHS)` mais il y a peut-être des cas que je n'ai pas en tête ;).

File: itou/scripts/tests.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (celine-m-s) [reply itou/scripts/tests.py:188]: Mon test ne passe plus si je supprime suspension4.
J'avoue que je ne pense pas que ce soit très utile de peaufiner pour un script qui ne tournera qu'une fois. Qu'en dis-tu ?

File: itou/scripts/management/commands/update_suspensions_end_at.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (rsebille) [reply itou/scripts/management/commands/update_suspensions_end_at.py:116]: +1 pour mettre une transaction sur le `handle()`, ça nous évitera le cas chelou et particulier où quelqu'un modifie une suspension au même moment (improbable je sais, mais les utilisateurs défient toujours les probabilités :grin:)

[the remark this answers is not in the document]
Comment (rsebille) [reply itou/scripts/management/commands/update_suspensions_end_at.py:116]: +1 pour mettre une transaction sur le `handle()`, ça nous évitera le cas chelou et particulier où quelqu'un modifie une suspension au même moment (improbable je sais, mais les utilisateurs défient toujours les probabilités :grin:)

**Comment by francoisfreitag** [comment]
Après un test manuel (en utilisant `--wet-run`), je constate que les dates des suspensions sont bien correctes. En revanche, la date de fin des PASS IAE reste après 2030.

```sql
itou=# SELECT end_at FROM approvals_approval ORDER BY end_at DESC LIMIT 20;
   end_at
------------
 2204-06-05
 2035-09-01
 2033-07-28
 2033-04-03
 2033-03-29
 2033-02-26
 2033-02-26
 2033-02-12
 2033-02-03
 2033-02-03
 2033-01-29
 2033-01-29
 2033-01-29
 2033-01-20
 2033-01-15
 2033-01-10
 2033-01-09
 2033-01-06
 2033-01-05
 2033-01-04
(20 rows)
```

**Review by rsebille — COMMENTED** [review]
@francoisfreitag Faut limiter au PASS qui ont été suspendu, car on fait des prolongations de 10 ans pour les CDI Inclusion et les contrats des 50 ans et plus (c'est peut-être les mêmes en fait :sweat_smile:).

File: itou/scripts/tests.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply itou/scripts/tests.py]: Ça me convient, c’est sûr que l’important pour ce script est la correction, et que les tests sont déjà un gros bonus.

**Comment by francoisfreitag** [comment]
> @Faut limiter au PASS qui ont été suspendu, car on fait des prolongations de 10 ans pour les CDI Inclusion et les contrats des 50 ans et plus

Effectivement, avec les filtres supplémentaires, c’est tout bon ! Merci

**Review by francoisfreitag — APPROVED** [review]

File: itou/scripts/management/commands/update_suspensions_end_at.py (hunk 6 above)
Comment (tonial) [reply itou/scripts/management/commands/update_suspensions_end_at.py]: J'ai changé la logique à ce niveau là en reprenant la première mise à jour de date de fin.

File: itou/scripts/tests.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (tonial) [reply itou/scripts/tests.py:188]: Je ne suis globalement pas convaincu de l'utilité du test vu qu'on peut l’exécuter en dry run pour vérifier le résultat.
Cela étant, comme il existait déjà, je l'ai adatpé à mes changements, et écrit à la main le CSV parce qu'en effet, la génération à partir des suspensions ne correspond pas à ce que l'on veut (avoir un PASS avec suspension traitées 2 fois.
